contracts/misc/UniswapV3OracleWrapper.sol

```diff
@@ -26,87 +28,251 @@ contract UniswapV3OracleWrapper is IAtomicPriceAggregator {
+    }
+
+    function getPendingFeeAmounts(
+        DataTypes.UinswapV3PositionData memory positionData
+    )
+        internal
+        view
+        returns (uint256 unclaimedTokenAFees, uint256 unclaimedTokenBFees)
+    {
+        IUniswapV3PoolState pool = IUniswapV3PoolState(
+            UNISWAP_V3_FACTORY.getPool(
+                positionData.tokenA,
+                positionData.tokenB,
+                positionData.fee
+            )
+        );
+        FeeParams memory feeParams;
+
+        (
+            ,
+            ,
+            feeParams.feeGrowthOutside0X128Lower,
+            feeParams.feeGrowthOutside1X128Lower,
+            ,
+            ,
+            ,
+
+        ) = pool.ticks(positionData.tickLower);
+        (
+            ,
+            ,
+            feeParams.feeGrowthOutside0X128Upper,
+            feeParams.feeGrowthOutside1X128Upper,
+            ,
+            ,
+            ,
+
+        ) = pool.ticks(positionData.tickUpper);
+
+        uint256 feeGrowthGlobal0X128 = pool.feeGrowthGlobal0X128();
+        uint256 feeGrowthGlobal1X128 = pool.feeGrowthGlobal1X128();
+
+        (, int24 tickCurrent, , , , , ) = pool.slot0();
+
+        unchecked {
+            // calculate fee growth below
+            uint256 feeGrowthBelow0X128;
+            uint256 feeGrowthBelow1X128;
+            if (tickCurrent >= positionData.tickLower) {
+                feeGrowthBelow0X128 = feeParams.feeGrowthOutside0X128Lower;
+                feeGrowthBelow1X128 = feeParams.feeGrowthOutside1X128Lower;
+            } else {
+                feeGrowthBelow0X128 =
+                    feeGrowthGlobal0X128 -
+                    feeParams.feeGrowthOutside0X128Lower;
+                feeGrowthBelow1X128 =
+                    feeGrowthGlobal1X128 -
+                    feeParams.feeGrowthOutside1X128Lower;
+            }
+
+            // calculate fee growth above
+            uint256 feeGrowthAbove0X128;
+            uint256 feeGrowthAbove1X128;
+            if (tickCurrent < positionData.tickUpper) {
+                feeGrowthAbove0X128 = feeParams.feeGrowthOutside0X128Upper;
+                feeGrowthAbove1X128 = feeParams.feeGrowthOutside1X128Upper;
+            } else {
+                feeGrowthAbove0X128 =
+                    feeGrowthGlobal0X128 -
+                    feeParams.feeGrowthOutside0X128Upper;
+                feeGrowthAbove1X128 =
+                    feeGrowthGlobal1X128 -
+                    feeParams.feeGrowthOutside1X128Upper;
+            }
+            uint256 feeGrowthInside0X128;
+            uint256 feeGrowthInside1X128;
+
+            feeGrowthInside0X128 =
+                feeGrowthGlobal0X128 -
+                feeGrowthBelow0X128 -
+                feeGrowthAbove0X128;
+            feeGrowthInside1X128 =
+                feeGrowthGlobal1X128 -
+                feeGrowthBelow1X128 -
+                feeGrowthAbove1X128;
+
+            unclaimedTokenAFees =
```

Comment:
we can use `FullMath.mulDiv` maybe to be more accurate

Reply:
yeah. i think so.